### PR TITLE
[WIP] [GSoC] Corrected the symbolic XFAIL tests in test_formal.py

### PR DESCRIPTION
**Pull request by arighnac**
#### Brief description of what is fixed or changed
Currently, there were some symbolic functions for which the `fps` of the functions could not be calculated.
For example :-
```
x, n = symbols('x n')
f = x**n*sin(x**2)
assert fps(f,x).truncate(8) = x**2*x**n - x**6*x**n/6 + O(x**(n + 8), x)
```
The above test would fail previously. This PR introduces the changes to correctly execute the symbolic `fps` tests.

#### Other comments
Please note that it is a work in progress.

#### Release Notes

* series
    * Added support for formal power series with symbolic `x**n` terms in `sympy.series.formal`

### CODE REVIEW

File: sympy/series/tests/test_formal.py [hunk 1]
```diff
@@ -395,21 +393,20 @@ def test_fps__symbolic():
...
+         (x - 2)**5*(x - 2)**n/1215 + O((x - 2)**(n + 6), (x, 2)))
 
     f = x**n*atan(x)
     assert fps(f, x, oo).truncate() == \
-        (-x**n/(5*x**5) + x**n/(3*x**3) + x**n*(pi/2 - 1/x) +
-         O(x**(n - 6), (x, oo)))
-
+        (-(1/x)**n/(5*x**5) + (1/x)**n/(3*x**3) + (1/x)**n*(pi/2 - 1/x) +
+         O((1/x)**(n + 6), (x, oo)))
+
```
Comment (oscargus) [inline sympy/series/tests/test_formal.py]: Here are the spaces causing the code quality test to fail..

Comment (ArighnaIITG) [reply sympy/series/tests/test_formal.py]: Done.

**Comment by oscargus** [comment]
I think that the release note should focus on the added functionality rather than the test. Something like
"Added support for formal power series with x**n terms".

File: sympy/series/formal.py [hunk 2]
```diff
@@ -812,6 +813,43 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
...
+                                symb = subt
+                                fps_sym = True
+                else:
+                    if not (isinstance(term, Pow) and term.exp.is_symbol):
+                        nterm *= term
+            nterms.append(nterm)
+        else:
+            nterms.append(t)
+
+    f = Add(*nterms)
```
Comment (leosartaj) [inline sympy/series/formal.py]: It will be helpful if you can add some comments explaining the not so obvious parts. It will help in maintaining it. I feel the code is a little complicated and could be simplified. Is it possible to extract some terms and use the existing algorithms on the remaining terms and maybe get the results back together? Is something like that possible?

File: sympy/series/formal.py [hunk 3]
```diff
@@ -824,8 +862,12 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
         return None
 
     ak = sequence(result[0], (k, result[2], oo))
-    xk = sequence(x**k, (k, 0, oo))
-    ind = result[1]
+    if fps_sym:
+        xk = sequence(x**k*x**symb, (k, 0, oo))
```
Comment (leosartaj) [inline sympy/series/formal.py]: It looks a bit too heuristic. Can we have a more generic approach here?

File: sympy/series/formal.py [hunk 4]
```diff
@@ -974,6 +1016,21 @@ def infinite(self):
...
+        for t in Add.make_args(ind):
+            if isinstance(t, Mul):
+                fterm, sterm = Mul.make_args(t)
+                if (isinstance(fterm, Pow) and fterm.exp.is_symbol) or \
+                 (isinstance(sterm, Pow) and sterm.exp.is_symbol):
+                    return S.One
+            else:
+                if isinstance(t, Pow) and t.exp.is_symbol:
+                    return S.One
+
```
Comment (leosartaj) [inline sympy/series/formal.py]: I believe there is already a function for this. I will see if I can look it up.

File: sympy/series/formal.py [hunk 5]
```diff
@@ -988,7 +1045,10 @@ def polynomial(self, n=6):
         terms = []
         for i, t in enumerate(self):
             xp = self._get_pow_x(t)
-            if xp >= n:
+            if xp.is_symbol:
+                terms.append(t)
+                continue
+            elif xp >= n:
```
Comment (leosartaj) [inline sympy/series/formal.py]: Can you add comments for the changes.

**Comment by leosartaj** [comment]
I have not fully tested it yet, but for something like this

```
from sympy import symbols, sin, fps
x, n = symbols('x n')
e = x**(n/2)*sin(x)
fps(e, x).truncate(10)
x - x**3/6 + x**5/120 - x**7/5040 + x**9/362880 + O(x**10)
```

Is this behaviour expected? Can you check this?

File: sympy/series/formal.py (hunk 2 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: Actually, the code is all about extracting the symbolic terms, and its various versions, like `x**(n-2)` or `x**(n/2)`, and then applying the rest of the functions (containing only function of x terms) to the existing algorithm. I will try to refactor the code, and add comments wherever necessary.

File: sympy/series/formal.py (hunk 5 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: Sure.

File: sympy/series/formal.py (hunk 3 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: I think that can't be changed. A heuristic approach simplifies things a lot here.

**Comment by ArighnaIITG** [comment]
@leosartaj , the above test case you mentioned was actually wrong. I modified the code, and now it is equipped to handle all symbolic functions, except the ones listed in the `XFAIL` block.

File: sympy/series/formal.py [hunk 6]
```diff
@@ -812,6 +813,46 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
...
+                            if sympify(res).is_integer:
+                                nterm *= x**res
+                            symb = subt
+                            fps_sym = True
+
+                    if isinstance(term.exp, Mul):
+                        symb = term.exp
+                        fps_sym = True
+            else:
+                if not (isinstance(term, Pow) and term.exp.is_symbol):
```
Comment (czgdp1807) [inline sympy/series/formal.py]: As far as I can figure out, control will reach `else` part only if, `isinstance(term, Pow)` is `False`. Now, `if not (isinstance(term, Pow) and term.exp.is_symbol)` can be interpreted as, `if not isinstance(term, Pow) or not term.exp.is_symbol` by DeMorgan's Laws and this will always be `True` as `not isinstance(term, Pow)` will always be `True`. So, overall, the check in L852 can be removed. Am I right?

File: sympy/series/formal.py [hunk 7]
```diff
@@ -974,6 +1020,17 @@ def infinite(self):
...
+        ind = self.ind
+        for t in Add.make_args(ind):
+            if isinstance(t, Mul):
+                for subt in Mul.make_args(t):
+                    if (isinstance(subt, Pow) and subt.exp.is_symbol):
+                        return S.One
+            else:
+                if isinstance(t, Pow):
+                    if isinstance(t.exp, Mul) or t.exp.is_symbol:
+                        return S.One
```
Comment (czgdp1807) [inline sympy/series/formal.py]: Can the conditions in L1030 and L1031 be written as, 
```
if isinstance(t, Pow) and (isinstance(t.exp, Mul) or t.exp.is_symbol):
``` 
I believe that nesting should be avoided wherever necessary. It helps in debugging using a debugger and enhances readability. If you follow some other principles then no issues.

File: sympy/series/formal.py (hunk 6 above)
Comment (leosartaj) [reply sympy/series/formal.py]: Hi @czgdp1807 thanks for analyzing this.

Hi @ArighnaIITG 

How about using something like this?

```
>>> from sympy import symbols, sin
>>> x, n = symbols('x n')
>>> e = x**n*sin(x**2)
>>> e.as_coeff_mul(x)
(1, (x**n, sin(x**2)))
```

I feel using something like this can greatly simplify it and not have you implement a custom term separation loop.

Comment (leosartaj) [reply sympy/series/formal.py]: On further thoughts,

You can continue this
```
>>> _, (t1, t2) = e.as_coeff_mul(x)
>>> fs = e.free_symbols
{x, n}
# for fps(e, x) free symbols will be fs - {x}, eg. here {n}
>>> t1.as_base_exp()
(x, n) # maybe presence of free symbols in exponential can be a good indication.
```

**Comment by ArighnaIITG** [comment]
@leosartaj A quite lot of changes have been made to make the code short. Kindly review them.

File: sympy/series/formal.py (hunk 7 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: Nestings have been removed. Thanks a lot @czgdp1807

File: sympy/series/formal.py [hunk 8]
```diff
@@ -972,8 +980,19 @@ def infinite(self):
 
         return self.ind + inf_sum
 
+    @property
+    def fps_sym(self):
+        """Returns whether the function contains symbolic terms"""
+        ind = self.ind
+        s = list(ind.free_symbols.difference({self.x}))
```
Comment (czgdp1807) [inline sympy/series/formal.py]: Can the conversion to `list` be avoided here as `set` also supports `len`?

File: sympy/series/formal.py [hunk 9]
```diff
@@ -810,6 +811,13 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
             return ak, xk, ind
         return None
 
+    syms = list(f.free_symbols.difference({x}))
+    symb = S.One
+    if len(syms) == S.One:
+        (t1, t2) = expand(f).as_independent(syms[0])
```
Comment (czgdp1807) [inline sympy/series/formal.py]: Is it possible to directly write, `(f, symb) = expand(f).as_independent(syms[0])` if `t1` and `t2` are not required anywhere in this file?

File: sympy/series/formal.py [hunk 10]
```diff
@@ -824,8 +832,8 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
         return None
 
     ak = sequence(result[0], (k, result[2], oo))
-    xk = sequence(x**k, (k, 0, oo))
-    ind = result[1]
+    xk = sequence(x**k*symb, (k, 0, oo))
+    ind = result[1]*symb
```
Comment (czgdp1807) [inline sympy/series/formal.py]: Please add spaces around `*`.

File: sympy/series/formal.py (hunk 8 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: True.

**Comment by ArighnaIITG** [comment]
ping @leosartaj . All the tests are passing now. I also included the test above.

File: sympy/series/formal.py [hunk 11]
```diff
@@ -810,6 +811,11 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
             return ak, xk, ind
         return None
 
+    syms = list(f.free_symbols.difference({x}))
```
Comment (leosartaj) [inline sympy/series/formal.py]: Can we have a ``free_symbols`` property instead? I am not sure but I think it should work in general. Can you check?

File: sympy/series/formal.py [hunk 12]
```diff
@@ -810,6 +811,11 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
...
+    symb = S.One
+    if len(syms) == S.One:
```
Comment (leosartaj) [inline sympy/series/formal.py]: What happens for something like ``x*(m+n)*sin(x**2)``?

File: sympy/series/formal.py [hunk 13]
```diff
@@ -972,8 +978,19 @@ def infinite(self):
...
         return self.ind + inf_sum
 
+    @property
+    def fps_sym(self):
+        """Returns whether the function contains symbolic terms"""
+        ind = self.ind
+        s = ind.free_symbols.difference({self.x})
+        if len(s) == S.One:
+            return True
+        return False
```
Comment (leosartaj) [inline sympy/series/formal.py]: Do we need this? I think having just a ``free_symbols`` property for ``FormalPowerSeries`` should suffice.

File: sympy/series/formal.py [hunk 14]
```diff
@@ -972,8 +978,19 @@ def infinite(self):
 
         return self.ind + inf_sum
 
+    @property
+    def fps_sym(self):
+        """Returns whether the function contains symbolic terms"""
+        ind = self.ind
+        s = ind.free_symbols.difference({self.x})
+        if len(s) == S.One:
```
Comment (leosartaj) [inline sympy/series/formal.py]: Again, What happens for something like x*(m+n)*sin(x**2)?

File: sympy/series/formal.py [hunk 15]
```diff
@@ -972,8 +978,19 @@ def infinite(self):
...
+            return True
+        return False
+
     def _get_pow_x(self, term):
         """Returns the power of x in a term."""
+        if self.fps_sym:
+            return S.One
```
Comment (leosartaj) [inline sympy/series/formal.py]: Why is this change needed? I am not very sure about this. This breaks the purpose of the function (behaves differently for a special case).

Comment (ArighnaIITG) [reply sympy/series/formal.py]: Suppose, for
`fps(x**n*exp(x),x).truncate(4) == x**n + x**n*x + x**n*x**2/2 + x**n*x**3/6 + O(x**4, x)`

Take any term, for eg. `x**n*x`. If you do a `as_base_exp()` operation on the term, we will get `xterm = x**n*x, pow_x = 1`. The problem comes if `x**n` term comes. `pow_x` then is `n`, a symbol. To generalize it, I kept return value as `S.One` for all symbolic fps.

File: sympy/series/formal.py (hunk 12 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: I had implemented it keeping one symbol in mind. I will try to change it.

File: sympy/series/formal.py (hunk 15 above)
Comment (leosartaj) [reply sympy/series/formal.py]: Can we avoid sending in the ``x**n`` term? It's a constant with respect to the algorithm.

File: sympy/series/formal.py (hunk 11 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: I didn't get you. Could you elaborate?

Comment (leosartaj) [reply sympy/series/formal.py]: https://docs.sympy.org/latest/modules/core.html#sympy.core.basic.Basic.free_symbols

File: sympy/series/formal.py (hunk 15 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: But in `polynomial` function of fps, the line `for i, t in enumerate(self):` will catch it. Are you telling we start `xk` from `(k, 1, oo)` instead of `(k, 0, oo)`, to get rid of the independent term?

File: sympy/series/formal.py (hunk 11 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: I have read that. But I couldn't understand which property you are mentioning? Can you tell that? There are no properties of `free_symbols` given in the above link.

File: sympy/series/formal.py (hunk 12 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: Done. @leosartaj

**Comment by ArighnaIITG** [comment]
ping @leosartaj . Code is now equipped to handle multiple symbols.

File: sympy/series/formal.py (hunk 11 above)
Comment (leosartaj) [reply sympy/series/formal.py]: > I have read that. But I couldn't understand which property you are mentioning? Can you tell that? There are no properties of `free_symbols` given in the above link.

Can you please recheck.

File: sympy/series/formal.py (hunk 15 above)
Comment (leosartaj) [reply sympy/series/formal.py]: Ideally, one would expect for the following thing to work.

```
>>> f = fps(sin(x), x)
>>> f._get_pow_x(x**(n+1))
>>> n + 1 # correct
>>> f._get_pow_x(x**n*x)
1 # wrong
```

Try using ``powsimp``. eg.

```
>>> powsimp(x**n*x, x)
x**(n+1)
```

File: sympy/series/formal.py [hunk 16]
```diff
@@ -1220,8 +1239,8 @@ def fps(f, x=None, x0=0, dir=1, hyper=True, order=4, rational=True, full=False):
     Examples
     ========
 
-    >>> from sympy import fps, O, ln, atan
-    >>> from sympy.abc import x
+    >>> from sympy import fps, O, ln, atan, sin
+    >>> from sympy.abc import x, n
```
Comment (leosartaj) [inline sympy/series/formal.py]: If you are importing this, why not add an example?

File: sympy/series/tests/test_formal.py [hunk 17]
```diff
@@ -377,38 +377,43 @@ def test_fps__logarithmic_singularity_fail():
...
+         (x - 2)**5*(x - 2)**n/1215 + O((x - 2)**(n + 6), (x, 2)))
+
+    f = x**(n - 2)*cos(x)
+    assert fps(f, x).truncate() == \
+        (x**n*x**(-2) - x**n/2 + x**2*x**n/24 - x**4*x**n/720 + O(x**(n + 6), x))
+
+    f = x**(n - 2)*sin(x) + x**n*exp(x)
+    assert fps(f, x).truncate() == \
+        (x**n/x + x**n + 5*x*x**n/6 + x**2*x**n/2 + 7*x**3*x**n/40 +
+         x**4*x**n/24 + 41*x**5*x**n/5040 + O(x**(n + 6), x))
```
Comment (leosartaj) [inline sympy/series/tests/test_formal.py]: Is it possible to display results like instead of ``x**4*x**n/24`` => ``x**(n+4)/24``. IMHO it might just be a simple`` powsimp`` call at the right place.

**Comment by ArighnaIITG** [comment]
ping @leosartaj

File: sympy/series/formal.py [hunk 18]
```diff
@@ -972,6 +980,11 @@ def infinite(self):
 
         return self.ind + inf_sum
 
+    @property
+    def free_symbols(self):
+        """Returns the free symbols of the formal power series"""
+        return self.function.free_symbols
```
Comment (leosartaj) [inline sympy/series/formal.py]: I checked we  already have this property implemented in ``SeriesBase``. Checkout the implementation to see how to implement ``free_symbols`` correctly. Remove this.

File: sympy/series/formal.py [hunk 19]
```diff
@@ -982,12 +995,16 @@ def _get_pow_x(self, term):
     def polynomial(self, n=6):
         """Truncated series as polynomial.
 
-        Returns series sexpansion of ``f`` upto order ``O(x**n)``
+        Returns series expansion of ``f`` upto order ``O(x**n)``
         as a polynomial(without ``O`` term).
         """
         terms = []
         for i, t in enumerate(self):
+            t = powsimp(t)
```
Comment (leosartaj) [inline sympy/series/formal.py]: Can this be moved inside ``_get_pow_x`` instead?

File: sympy/series/formal.py [hunk 20]
```diff
@@ -982,12 +995,16 @@ def _get_pow_x(self, term):
...
             xp = self._get_pow_x(t)
+            sym = self.free_symbols
```
Comment (leosartaj) [inline sympy/series/formal.py]: Probably don't need this in a loop.

File: sympy/series/formal.py [hunk 21]
```diff
@@ -1028,6 +1045,9 @@ def _eval_term(self, pt):
             ind = S.Zero
             for t in Add.make_args(self.ind):
                 pow_x = self._get_pow_x(t)
+                sym = self.free_symbols
```
Comment (leosartaj) [inline sympy/series/formal.py]: Probably don't need this in a loop.

File: sympy/series/formal.py [hunk 22]
```diff
@@ -810,6 +812,12 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
             return ak, xk, ind
         return None
 
+    syms = f.free_symbols.difference({x})
+    (f, symb) = expand(f).as_independent(*syms)
+    if symb is S.Zero:
+        symb = S.One
+    symb = powsimp(symb)
```
Comment (leosartaj) [inline sympy/series/formal.py]: Add a comment or two.

File: sympy/series/formal.py [hunk 23]
```diff
@@ -824,8 +832,8 @@ def _compute_fps(f, x, x0, dir, hyper, order, rational, full):
         return None
 
     ak = sequence(result[0], (k, result[2], oo))
-    xk = sequence(x**k, (k, 0, oo))
-    ind = result[1]
+    xk = sequence(x**k * symb, (k, 0, oo))
```
Comment (leosartaj) [inline sympy/series/formal.py]: Probably if you do a ``powsimp`` here 
```
xk_formula = powsimp(x**k * symb)
xk = sequence(xk_formula, (k, 0, oo))
```

This is will make it more easier to read as well. Probably you might be able to drop some ``powsimp`` calls.

**Comment by leosartaj** [comment]
General Comments,

First of all good work on the progress. Make sure to checkout https://github.com/sympy/sympy/wiki/Development-workflow#writing-commit-messages for learning to write better commit messages. In general as discussed before, try to go through this guide thoroughly.


>Release Notes
>series
>Added support for formal power series with symbolic x**n terms in sympy.series.tests.formal

``sympy.series.formal``

File: sympy/series/formal.py (hunk 19 above)
Comment (ArighnaIITG) [reply sympy/series/formal.py]: I don't think we need this anymore.

**Comment by ArighnaIITG** [comment]
@leosartaj , I have pushed in the changes. I will surely try to pay attention to the commit messages from the next time. You can verify it, and we can wait for the Travis tests to pass.

**Comment by leosartaj** [comment]
> You can verify it, and we can wait for the Travis tests to pass.

Please ping me once the tests pass.

File: sympy/series/formal.py [hunk 24]
```diff
@@ -972,6 +982,20 @@ def infinite(self):
 
         return self.ind + inf_sum
 
+    @property
+    def variables(self):
+        """Returns a tuple of variables that are bounded"""
+        return ()
```
Comment (leosartaj) [inline sympy/series/formal.py]: Not needed.

File: sympy/series/formal.py [hunk 25]
```diff
@@ -972,6 +982,20 @@ def infinite(self):
...
+
+    @property
+    def free_symbols(self):
+        """
+        This method returns the symbols in the object, excluding those
+        that take on a specific value (i.e. the dummy symbols).
+        """
+        return (set(j for i in self.args for j in i.free_symbols)
+                .difference(self.variables))
```
Comment (leosartaj) [inline sympy/series/formal.py]: Why re-implementing this? It's already part of ``SeriesBase``. Note ``FormalPowerSeries`` inherits from it.

Comment (ArighnaIITG) [reply sympy/series/formal.py]: Sorry. I kept that for checking something. Forgot to remove it. Thanks @leosartaj

**Comment by ArighnaIITG** [comment]
I have removed the `free_symbols` property from `FormalPowerSeries` class, since it is already being inherited.

**Comment by ArighnaIITG** [comment]
ping @leosartaj . All the tests have successfully passed. !!

**Comment by leosartaj** [comment]
> ping @leosartaj . All the tests have successfully passed. !!

Great! This looks good to me. I think it's mergeable now.

Hi @asmeurer @moorepants ,

I know that we do not encourage rebasing, but I see there are 19 commits for a comparatively simple change. What will be the best way to go about?

**Comment by leosartaj** [comment]
Hi @ArighnaIITG 

This is ready to be merged.  Good work! 

PS: Try to make more meaningful commits for subsequent PRs.
